src/indexable/sampler/batch_sampler.rs

```diff
@@ -97,7 +97,22 @@ where
         }
         None
     }
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (lower, upper) = self.sampler.size_hint();
+        if self.drop_last {
+            let lower = lower / self.batch_size;
+            let upper = upper.map(|upper| upper / self.batch_size);
+            (lower, upper)
+        } else {
+            let lower = (lower + self.batch_size - 1) / self.batch_size;
+            let upper = upper.map(|upper| (upper + self.batch_size - 1) / self.batch_size);
```

Comment:
As we are in the case of we know exactly the size of the remaining elements, I think we can just do
```rust
let upper = Some(lower);
```
This invariant is [aggressively checked](https://doc.rust-lang.org/stable/src/core/iter/traits/exact_size.rs.html#122) in the default implementation of `ExactSizeIterator` of the standard library, so I think we can also go this way.

Reply:
Indeed you're right but that means 
the iterator embedeed in BatchSampler has always an ExactSize ? 
```rust
impl<I> Iterator for BatchIterator<I>
where
    I: Iterator<Item = usize>,
```
I'm afraid that one day someone create a BatchIterator with an iterator that do not respect 
```rust 
let upper = Some(lower); // ie ExactSizeIterator definition
``` 
But if you think we are good to go I'll change it ;)

Reply:
And same for iterable dataset: 
```rust
impl<D, C> IntoIterator for DataLoader<D, C>
where
    D: IntoIterator,
    D::IntoIter: ExactSizeIterator,
    C: Collate<<D as IntoIterator>::Item>,
```

Reply:
That would be fine because you already add a where clause that specifies we implement `ExactSizeIterator` only if the underlying sampler also implement `ExactSizeIterator`.

```rust
impl<I> ExactSizeIterator for BatchIterator<I> where I: Iterator<Item = usize> + ExactSizeIterator {}
```

So if the sampler doesn't implement  `ExactSizeIterator`, neither we do, so we don't have to guarantee that the return of `size_hint` is true.

Reply:
Ok nice 👍  
Everything should be ok on the CI then.